### ISSUE TITLE
4.3's contextual typing of unioned signature breaks @types/linkify-it

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#42941]
### 💻 Code

```ts
// @strict: true
type Rule = {
    boomhauer(mumble: string, grumble = true): void
    normalize(region: "eastern" | "north" | "hill country"): string
}

declare function quickInfo(rule: string | Rule): void

quickInfo({
    boomhauer(msg) {
        console.log(msg)
    },
    normalize(region) {
        return region
    }
})
```

### 🙁 Actual behavior

No implicit any error on the parameter `region`.

### 🙂 Expected behavior

No error, and `region: string`.

This is an expected change of the recent change to contextual typing of unioned signatures as far as I can tell. If it is, then the fix for this bug needs to be a workaround for contextual typing in `@types/linkify-it` on DT.

**Comment by weswigham** [comment]
Yes, I referenced this in the PR, included it as a test case, and explained the issue in the design meeting - since the overloaded members of the union (ie, `String.normalize`) used to produce no signatures, we'd only pull signatures from the other union member. Now we recognize the `normalize` signature could come from either union member and, in fact, has conflicting types. We decided this was an OK break, since you'd already have gotten this same error if, eg, you removed one of the overloads from `String.normalize` in the lib (which would then have allowed the signatures to be compared under our prior logic, and recognized as inequal), and it makes no sense that adding an overload to a conceptually unrelated thing had the potential to change this contextual type anyway.
